python/ray/data/_internal/block_batching/util.py

```diff
@@ -1,15 +1,201 @@
 import logging
 import queue
 import threading
-from typing import Callable, Iterator, TypeVar
+import sys
+from typing import Any, Callable, Iterator, List, Optional, Tuple, TypeVar, Union
+
+import ray
+from ray.actor import ActorHandle
+from ray.types import ObjectRef
+from ray.data.block import Block, BlockAccessor, DataBatch
+from ray.data._internal.batcher import Batcher, ShufflingBatcher
+from ray.data._internal.block_batching.interfaces import (
+    Batch,
+    CollatedBatch,
+    BlockPrefetcher,
+)
+from ray.data._internal.memory_tracing import trace_deallocation
+from ray.data._internal.stats import DatasetPipelineStats, DatasetStats
+from ray.util.scheduling_strategies import NodeAffinitySchedulingStrategy
 
 T = TypeVar("T")
 U = TypeVar("U")
 
 logger = logging.getLogger(__name__)
 
+if sys.version_info >= (3, 7):
+    from contextlib import nullcontext
+else:
+    from contextlib import contextmanager
+
+    @contextmanager
+    def nullcontext(enter_result=None):
+        yield enter_result
+
+
+def _calculate_ref_hits(refs: List[ObjectRef[Any]]) -> Tuple[int, int, int]:
+    """Given a list of object references, returns how many are already on the local
+    node, how many require fetching from another node, and how many have unknown
+    locations."""
+    current_node_id = ray.get_runtime_context().get_node_id()
+
+    locs = ray.experimental.get_object_locations(refs)
+    nodes: List[List[str]] = [loc["node_ids"] for loc in locs.values()]
+    hits = sum(current_node_id in node_ids for node_ids in nodes)
+    unknowns = sum(1 for node_ids in nodes if not node_ids)
+    misses = len(nodes) - hits - unknowns
+    return hits, misses, unknowns
+
+
+def resolve_block_refs(
+    block_ref_iter: Iterator[ObjectRef[Block]],
+    eager_free: bool = False,
+    stats: Optional[Union[DatasetStats, DatasetPipelineStats]] = None,
+) -> Iterator[Block]:
+    """Resolves the block references for each logical batch.
+
+    Args:
+        block_ref_iter: An iterator over block object references.
+        eager_free: Whether to eagerly free the object reference from the object store.
+        stats: An optional stats object to recording block hits and misses.
+    """
+    hits = 0
+    misses = 0
+    unknowns = 0
+
+    for block_ref in block_ref_iter:
+        current_hit, current_miss, current_unknown = _calculate_ref_hits([block_ref])
+        hits += current_hit
+        misses += current_miss
+        unknowns += current_unknown
+
+        # TODO(amogkam): Optimized further by batching multiple references in a single
+        # `ray.get()` call.
+        block = ray.get(block_ref)
```

Comment:
Same here, this should have the appropriate stats instrumentation, right?
```suggestion
        with stats.iter_get_s.timer() if stats else nullcontext():
            block = ray.get(block_ref)
```

Reply:
Oops, had this in the third PR and forgot to add it to this one. Will update.

Reply:
updated